### ISSUE TITLE
[R] SKaiNET vs KotlinDL

### ISSUE THREAD

**Issue by michalharakal** [sk-ai-net/skainet#215]
KotlinDL is an official JetBrains project (https://github.com/Kotlin/kotlindl) that shares some common goals and technologies. This task should help decision-makers evaluate various criteria to choose which option should be used for which use cases, while understanding the pros and cons.

**Comment by michalharakal** [comment]
**Title:** Select appropriate Kotlin-based ML framework (KotlinDL vs SKaiNET)

**As a**
technical decision maker (e.g., Lead Developer / Architect / CTO),

**I want to**
evaluate and compare two Kotlin-based ML frameworks (KotlinDL and SKaiNET),

**so that**
I can select the right framework for our current and future products, based on vision, goals, technology fit, maturity, capabilities, and team availability.

---

## 2. Business Value

* Reduce **technical and strategic risk** when committing to an ML framework.
* Ensure **alignment with product vision** (e.g., mobile, backend, experimentation, education).
* Optimize **developer productivity and hiring** by choosing a framework that fits team skills and availability.
* Avoid **rewrite costs** by understanding when one framework might need to be replaced or complemented by another.

---

## 3. Scope

This user story covers:

* Evaluating **KotlinDL** and **SKaiNET** against defined criteria.
* Documenting **when to use which framework** (decision rules, not just a one-time choice).
* Producing **decision artifacts** (comparison sheet, risk assessment, recommendation).

Out of scope:

* Implementing the full production ML pipeline.
* Benchmarking against non-Kotlin ML frameworks (e.g., PyTorch, TensorFlow Python) beyond high-level notes.

---

## 4. What to Consider & What to Check

Use the following sections as a **checklist**.

### 4.1 Vision Alignment

**Consider:**

* What is the **long-term product vision**?

  * Mobile-first? Backend services? Educational tools? Internal research?
* Does each framework align with:

  * Need for **on-device inference** (Android)?
  * Need for **server-side ML** in a Kotlin backend?
  * Need for **simple educational/experimental models**?

**Check:**

* Is there a **clear description** of each framework’s intended purpose and use cases in their docs/readme?
* Do their examples match your **planned use cases**?

---

### 4.2 Product & Technical Goals

**Consider:**

* What are the **primary ML tasks**?

  * Vision (image classification, object detection)
  * Tabular data
  * Time series
  * NLP (if relevant)
* What are your **performance requirements**?

  * Latency thresholds
  * Throughput
  * Model size constraints (e.g., mobile footprint)

**Check:**

* Whether each framework:

  * Supports your **core ML tasks** out of the box.
  * Offers **pretrained models** or only DIY architectures.
  * Has **APIs for deployment** in your target environment (Android app, Ktor backend, Spring Boot, etc.).

---

### 4.3 Targets & KPIs

**Consider:**

What will define **success** of adopting a framework?

Examples:

* Time to first working prototype.
* Time to integrate with existing Kotlin codebase.
* Model inference latency on target hardware.
* Number of critical bugs encountered during pilot.

**Check:**

* Define **3–5 measurable KPIs** and ensure both frameworks are evaluated against them in a PoC (even minimal).

---

### 4.4 Audience & Usage Patterns

**Consider:**

* Who will mainly use the framework?

  * Backend engineers with limited ML background
  * ML engineers who are new to Kotlin
  * Students/trainees (education)
* Will it be:

  * Used mostly by **one team**, or
  * Shared across **multiple teams and projects**?

**Check:**

* How approachable is the **API and documentation** for your primary audience?
* Are there **tutorials, sample projects, and guides** that fit your audience level?

---

### 4.5 Technology & Architecture

**Consider:**

* Runtime environments:

  * JVM server
  * Android
  * Multiplatform?
* Dependency expectations:

  * Does the framework rely on **TensorFlow, native libs, or pure Kotlin**?
* Integration:

  * How easily can it plug into your **build system (Gradle)**, **CI/CD**, and **logging/monitoring** stack?

**Check:**

* Compatibility with:

  * Your **current Kotlin version** and JVM target.
  * Your **Android minSdk** (if mobile).
* Whether the framework:

  * Integrates easily with your **existing modules**.
  * Imposes any **platform-specific constraints** (e.g., native libs, ABI issues).

---

### 4.6 Maturity & Development Status

**Consider:**

* How **active** is the project?
* How **stable** are APIs? Frequent breaking changes or relatively stable?

**Check:**

* Last commit date.
* Number of contributors and maintainers.
* Release cadence (tags, versions, changelogs).
* Open issues vs closed issues (trend, not exact numbers).
* Presence of **versioning policy** (e.g., semantic versioning).

---

### 4.7 Capabilities

**Consider:**

* Supported model types:

  * Feedforward, CNNs, RNNs/LSTMs, transformers?
* Training vs inference:

  * Does it support **training**, **inference only**, or both?
* Utilities:

  * Dataset loaders, preprocessing, augmentation, serialization, export/import of models.

**Check:**

* Compare **feature coverage**:

  * Vision tools (image loading, augmentation, pretrained models).
  * Optimizers and loss functions.
  * GPU/CPU support details.
* Existing **examples**:

  * Do they cover similar complexity to your planned use case?

---

### 4.8 Developer Availability & Skills

**Consider:**

* Skill profile of your team:

  * Kotlin-strong, ML-weak?
  * ML-strong, Kotlin-weak?
* How hard will it be to:

  * Onboard new developers?
  * Hire or contract help?

**Check:**

* Availability of:

  * Tutorials, blog posts, talks, Q&A (e.g., Stack Overflow).
  * Community channels: Slack, Discord, GitHub Discussions, etc.
* Internal assessment:

  * Can 1–2 devs build a **minimal PoC** in each framework within a sprint?

---

### 4.9 Risk, Support & Governance

**Consider:**

* Is there a **company or institution** backing the framework?
* What happens if the framework becomes **unmaintained**?

**Check:**

* License compatibility with your product (e.g., Apache 2.0, MIT).
* Security posture: does it depend on risky native binaries or unmaintained dependencies?
* Migration path:

  * Can models be exported/imported in standard formats (e.g., ONNX, TensorFlow saved models)?

---

## 5. Acceptance Criteria

The user story is **done** when all the following are true:

1. **Documented Comparison**

   * A **written comparison document** exists (e.g., Confluence/Notion/MD file) that:

     * Lists **KotlinDL** and **SKaiNET**.
     * Compares them across:

       * Vision alignment
       * Product & technical goals and KPIs
       * Audience suitability
       * Technology & integration
       * Maturity & development status
       * Capabilities
       * Developer availability
       * Risks and licensing

2. **Proof-of-Concept Implementations**

   * For a **representative use case** (e.g., simple image classifier or tabular predictor), there is:

     * One **minimal PoC** implemented using KotlinDL.
     * One **minimal PoC** implemented using SKaiNET.
   * Each PoC:

     * Runs in your **target environment** (e.g., JVM service or Android app).
     * Logs **basic performance metrics** (latency, inference time, build size where relevant).

3. **Measured Evaluation**

   * KPIs defined earlier (e.g., time to prototype, latency, dev effort) are:

     * Explicitly measured or **estimated and justified** for both frameworks.
     * Captured in a **simple scorecard or decision matrix**.

4. **Decision & “When to Use What” Rules**

   * There is a **clear recommendation** for:

     * Primary/default framework to use going forward.
     * Situations where the **other framework is preferred** (e.g., education vs production).
   * The decision document includes simple rules, e.g.:

     * “Use Framework A when building production-grade Android vision features.”
     * “Use Framework B for teaching, quick prototypes, or lightweight experiments.”
   * These rules are **reviewed and agreed** by:

     * At least one lead engineer
     * The product/tech owner or architect

5. **Risk & Re-Evaluation Plan**

   * Identified **key risks** (e.g., project abandonment, missing features).
   * Defined **triggers for re-evaluation**, for example:

     * Major breaking change in chosen framework.
     * Project becomes inactive (no commits for N months).
     * New framework or major competitor appears.
   * Added a **note in the tech strategy / architecture decision log (ADR)** summarizing:

     * Decision made.
     * Rationale.
     * Date and owner.
     * Re-evaluation conditions.

6. **Onboarding & Knowledge Sharing**

   * At least one **short internal doc or README** describes:

     * How to set up the chosen framework.
     * How to run the PoC example.
   * At least one **knowledge-sharing session** (demo / brown bag) is held or documented:

     * Showing results from both frameworks.
     * Explaining why the decision was made.

---

## 6. Optional: Gherkin-Style Acceptance Examples

```gherkin
Scenario: Compare KotlinDL and SKaiNET
  Given two Kotlin-based ML frameworks: KotlinDL and SKaiNET
  When we implement the same ML use case as a PoC in both frameworks
  And we measure implementation effort, performance, and integration complexity
  Then we have a documented comparison across agreed evaluation criteria
  And we can clearly state which framework we will use in which types of projects.

Scenario: Decision is traceable
  Given the comparison document and PoC results
  When a new engineer joins the team
  Then they can understand in under 30 minutes
  Why KotlinDL or SKaiNET was chosen for production
  And in which cases the alternative should be considered.

Scenario: Re-evaluation trigger
  Given that we adopted one framework as the default
  When its development activity ceases or a critical limitation is hit
  Then we have a clear, documented trigger and process
  To re-evaluate Kotlin-based ML frameworks (including the previously rejected one).
```

**Comment by michalharakal** [comment]
## Feature Set
### SKaiNET

* Offers a Kotlin DSL for defining neural network architectures in code. For example, it supports layers like Conv2D, pooling, and Dense defined in a neat Kotlin DSL block. Common deep learning layers such as Dropout and Batch Normalization are built-in. 
* includes a data API with dataset loaders (e.g. an MNIST loader and JSON dataset support) for feeding training data. There’s even initial support for loading pre-trained models in GGUF format. 
* Under the hood, SKaiNET implements its own tensor operations engine – using the Java SIMD Vector API for CPU acceleration of math ops like matrix multiplication and convolution

### KotlinDL: 

* Provides both Sequential and Functional API styles (very similar to Keras) for constructing models
* It includes a wide range of predefined layers (Conv2D, pooling, Dense, Flatten, activations, etc.) and supports high-level training workflows: compile the model with an optimizer/loss, then fit() on a dataset, evaluate, etc
* KotlinDL makes it easy to use pre-trained models: it can import models from Keras (SavedModel or HDF5) and from ONNX for *inference*
* In fact, KotlinDL comes with a **ModelHub** that provides dozens of ready-to-use models – for example, popular image classifiers (ResNet, MobileNet, EfficientNet), object detectors (EfficientDet), pose detection models (MoveNet), etc., which can be downloaded and used or fine-tuned in Kotlin
* The library also includes a DSL for image preprocessing pipelines and supports special task-specific APIs (e.g. a high-level API for pose detection was introduced in v0.4)

**Comment by michalharakal** [comment]
Instead of comparison, let's think out hot to use SKaiNET as underlaying engine replacing outdated TF and ONNX.

See #290 for details
